sdks/java/extensions/ml/src/main/java/org/apache/beam/sdk/extensions/ml/AnnotateImages.java

```diff
@@ -0,0 +1,209 @@
+   */
+  @Override
+  public PCollection<List<AnnotateImageResponse>> expand(PCollection<T> input) {
+    ParDo.SingleOutput<T, AnnotateImageRequest> inputToRequestMapper;
+    if (contextSideInput != null) {
+      inputToRequestMapper =
+          ParDo.of(new MapInputToRequest(contextSideInput)).withSideInputs(contextSideInput);
+    } else {
+      inputToRequestMapper = ParDo.of(new MapInputToRequest(null));
+    }
+    return input
+        .apply(inputToRequestMapper)
+        .apply(ParDo.of(new AssignRandomKeys()))
+        .apply(GroupIntoBatches.ofSize(batchSize))
+        .apply(ParDo.of(new ExtractValues()))
+        .apply(ParDo.of(new PerformImageAnnotation()));
+  }
+
+  /**
+   * Input type to {@link AnnotateImageRequest} mapper. Needs to be implemented by child classes
+   *
+   * @param input Input element.
+   * @param ctx optional image context.
+   * @return A valid {@link AnnotateImageRequest} object.
+   */
+  public abstract AnnotateImageRequest mapToRequest(T input, ImageContext ctx);
+
+  /**
+   * The {@link DoFn} performing the calls to Cloud Vision API. Input PCollection contains lists of
+   * {@link AnnotateImageRequest}s ready for batching.
+   */
+  public static class PerformImageAnnotation
+      extends DoFn<List<AnnotateImageRequest>, List<AnnotateImageResponse>> {
+
+    private ImageAnnotatorClient imageAnnotatorClient;
+
+    public PerformImageAnnotation() {}
+
+    /**
+     * Parametrized constructor to make mock injection easier in testing.
+     *
+     * @param imageAnnotatorClient
+     */
+    public PerformImageAnnotation(ImageAnnotatorClient imageAnnotatorClient) {
+      this.imageAnnotatorClient = imageAnnotatorClient;
+    }
+
+    @StartBundle
+    public void startBundle() throws IOException {
+      imageAnnotatorClient = ImageAnnotatorClient.create();
+    }
+
+    @Teardown
+    public void teardown() {
+      imageAnnotatorClient.close();
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      context.output(getResponse(context.element()));
+    }
+
+    /**
+     * Performs the call itself. Default access for testing.
+     *
+     * @param requests request list.
+     * @return response list.
+     */
+    List<AnnotateImageResponse> getResponse(List<AnnotateImageRequest> requests) {
+      BatchAnnotateImagesResponse batchAnnotateImagesResponse =
+          imageAnnotatorClient.batchAnnotateImages(requests);
+      return batchAnnotateImagesResponse.getResponsesList();
+    }
+  }
+
+  /** A transform that converts input elements to {@link KV}s for grouping. */
+  private static class AssignRandomKeys
+      extends DoFn<AnnotateImageRequest, KV<Long, AnnotateImageRequest>> {
+    private Random random;
+
+    @Setup
+    public void setup() {
+      random = new Random();
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      context.output(KV.of(random.nextLong(), context.element()));
```

Comment:
You're right. Java SDK doesn't really have a transform that batches the elements in bundles of set size regardless of key - do you think we can omit the batching and process all the elements serially?

Reply:
If batching is to reduce the number of calls to the annotation API, what about using a collection in the DoFn instead to batch up requests?

Assigning random keys and grouping could possibly increase the transform parallelism depending on the source, the data distribution, the operations in the upstream graph, and the runner capabilities but at this point I think the user would be most capable of making that decision. Grouping requires a shuffle and that is an expensive operation that should be avoided if possible.

Reply:
I tried creating my own batching DoFn and discovered that input elements need to be KVs anyway - to get around it I set all keys for input KVs to zero and figured that I can just reuse `GroupIntoBatches` since it does exactly the same thing as my custom DoFn anyway.

WDYT about this approach?

Reply:
+1 on using GroupIntoBatches

Reply:
@lukecwik Are you OK with assigning the same keys to all elements just to use `GroupIntoBatches` without friction?

Reply:
Please use a random assignment using a fixed number of keys for now. There have been some discussions in Google about making GroupIntoBatches work efficiently even if there is only one key but that project will take months before being realized.

Reply:
Won't having a single key result in all the records going through a single worker thread after the GroupIntoBatches?

@lukecwik could you recommend a fixed number of keys? I think this is hard to get right since it will depend on the input size and desired parallelism of worker threads.

Reply:
I can't as this is IO dependent and I don't know what the QPS expectations for a pipeline for this API are. For example 50 made sense for BigQuery but it was doing 1000s of QPS.

Reply:
OK so I see two reasonable options then,

1. determine the QPS expectations or limits of using this transform in a pipeline and select a default number of keys based on this (with a comment to explain the logic and an option to override)
2. parameterize the number of keys with no default

@mwalenia either is fine with me.

Reply:
The default limit of QPS is 30 (1800 requests per minute). I don't know what the default number of keys should be based on that - @lukecwik can you help me here?

I'll start with parametrizing the number with no default. Thanks!

Reply:
@lukecwik What do you think the number of keys should be, based on the default limit of QPS being 30?

Reply:
A good rule of thumb would be to follow `average latency / QPS ~= # keys`

Reply:
I think this will be best left for the user, as latency can change and the limit can be changed by GCP project owners.

Reply:
@lukecwik if you're OK with leaving it without a default, I think we can merge this.

Reply:
Since the default limit is 30QPS but the latency will depend a lot on the size of the input let's be cautious, assume smaller sized input data, and use something small like 5 keys.